File: client/lib/domain/entities/message.dart
import 'package:equatable/equatable.dart';

enum MessageRole { user, assistant }

class Message extends Equatable {
  final String id;
  final String content;
  final MessageRole role;
  final DateTime createdAt;

  const Message({
    required this.id,
    required this.content,
    required this.role,
    required this.createdAt,
  });

  Map<String, dynamic> toJson() => {
    'role': role == MessageRole.user ? 'user' : 'assistant',
    'content': content,
  };

  @override
  List<Object?> get props => [id, content, role, createdAt];
}

File: internal/domain/message.go
package domain

type MessageRole string

const (
	MessageRoleSystem    MessageRole = "system"
	MessageRoleUser      MessageRole = "user"
	MessageRoleAssistant MessageRole = "assistant"
)

func (m *Message) ToMap() map[string]interface{} {
	return map[string]interface{}{
		"role":    string(m.Role),
		"content": m.Content,
	}
}

func FromProtoRole(role string) MessageRole {
	switch role {
	case "system":
		return MessageRoleSystem
	case "user":
		return MessageRoleUser
	case "assistant":
		return MessageRoleAssistant
	default:
		return MessageRoleUser
	}
}

func ToProtoRole(role MessageRole) string {
	return string(role)
}
